### PR TITLE
Rules syntax highlighting & hover & code lens

### PR DESCRIPTION
**Pull request by pieterbos**
Adds to rules:

- syntax highlighting
- code lens for paths
- hovers for paths

Note that the commit message says this does not work for rules in template overlays. However, these cannot contain rules, so not an issue :)

Looks like this:
Screenshot 2021-02-18 at 12 37 52

### CODE REVIEW

File: src/main/java/com/nedap/openehr/lsp/symbolextractor/SymbolInformationExtractingListener.java [hunk 1]
```diff
@@ -95,8 +106,18 @@ public void exitArchetype(AdlParser.ArchetypeContext ctx) {
...
+        setArchetypeId(ctx.ARCHETYPE_HRID());
+    }
+
+    private void setArchetypeId(TerminalNode terminalNode) {
+        this.archetypeId = terminalNode.getText();
+        this.currentOverlayid = archetypeId;
     }
+
+    private void setOverlayId(TerminalNode terminalNode) {
+        this.currentOverlayid = archetypeId;
```
Comment (VeraPrinsen) [inline src/main/java/com/nedap/openehr/lsp/symbolextractor/SymbolInformationExtractingListener.java:118]: I suspect the currentOverlayId must be get from the terminalNode (which is unused at the moment)? ArchetypeId could possibly be null

File: src/main/java/com/nedap/openehr/lsp/repository/BroadcastingArchetypeRepository.java [hunk 2]
```diff
@@ -432,7 +435,35 @@ public ADL2TextDocumentService getTextDocumentService() {
...
+        List<CodeLens> result = new ArrayList<>();
+        for(ArchetypePathReference reference:documentInformation.getModelReferences()) {
+            if(reference.getArchetypeId() == null) {
+                continue;
+            }
+            try {
+                ValidationResult validationResult = getValidationResult(documentInformation.getArchetypeId());
+                if (validationResult != null) {
+                    PathUtils.createCodeLenses(result, reference, validationResult);
+                } else {
```
Comment (VeraPrinsen) [inline src/main/java/com/nedap/openehr/lsp/repository/BroadcastingArchetypeRepository.java:459]: Empty else statement

File: src/main/java/com/nedap/openehr/lsp/paths/PathUtils.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                String text = content + partial.getRemainingQuery().stream().map(PathSegment::toString).collect(Collectors.joining("/"));
+                String extraText = description;
+                extraText += "\n\nIn Archetype " + flattened.getDefinition().getTerm().getText() + " (" + reference.getArchetypeId() +")";
+                if(reference.getExtraInformation() != null) {
+                    text += " " + reference.getExtraInformation();
+                }
+
+                //result.add(new CodeLens(reference.getRange(), new Command(text, ADL2TextDocumentService.SHOW_INFO_COMMAND, Lists.newArrayList(extraText)), null));
+                result.add(new CodeLens(reference.getRange(), new Command(text, ADL2TextDocumentService.SHOW_INFO_COMMAND, Lists.newArrayList(extraText)), null));
+            } else {
```
Comment (VeraPrinsen) [inline src/main/java/com/nedap/openehr/lsp/paths/PathUtils.java:51]: Empty else statements and commented code

**Review by VeraPrinsen — CHANGES_REQUESTED** [review]
I suspect the setOverlayId method in the SymbolInformationExtractingListener needs a change.

Furthermore I saw some empty statements or commented code, but that's just cleanup.

Lastly this is totally new code for me so I really miss some documentation in general, but maybe that is not worth it at the moment? But because this is about visuals I would approve if I know that the setOverlayId method is right.

File: src/main/java/com/nedap/openehr/lsp/symbolextractor/SymbolInformationExtractingListener.java (hunk 1 above)
Comment (pieterbos) [reply src/main/java/com/nedap/openehr/lsp/symbolextractor/SymbolInformationExtractingListener.java:118]: good catch, will change!

File: src/main/java/com/nedap/openehr/lsp/paths/PathUtils.java (hunk 3 above)
Comment (pieterbos) [reply src/main/java/com/nedap/openehr/lsp/paths/PathUtils.java:51]: The results.add I will remove. I think the empty else and potential hover add something: it indicates the easy possibility of adding that feature if we decide it's actually better than not having it.

File: src/main/java/com/nedap/openehr/lsp/repository/BroadcastingArchetypeRepository.java (hunk 2 above)
Comment (pieterbos) [reply src/main/java/com/nedap/openehr/lsp/repository/BroadcastingArchetypeRepository.java:459]: will remove

**Review by VeraPrinsen — APPROVED** [review]

**Pull request merged**